### ISSUE TITLE
TearDown is executed BEFORE User Attribute AfterTest(ITest test) starting from Nunit 3.7.0

### ISSUE THREAD

**Issue by andrewlaser** [nunit/nunit#2329]
In Nunit 3.6.1 and older there was the following sequence:

1. testcase itself
2. custom attribute AfterTest(ITest test)
3. [TearDown] TestFixtureTearDown()

Since v.3.7.0  the sequence was broken (now it is 1, 3, 2) , causing our tests failure, because TearDown deletes data used by custom AfterTest

**TestCases.cs**
    [TestFixture]
    [Parallelizable]
    public class LoginCases : TestBase
    {
        [Test]
        public void MyTest()
        {
........
         }

**TestBase.cs**
namespace RS.Tests
{
    [TestFixture]
    [SoftAssertions]
    public class TestBase
{
...
        /// <summary>
        /// TearDown that runs AFTER each test
        /// </summary>
        [TearDown]
        public void TestFixtureTearDown()
        {
...
            Close(); //delete Webdriver
        }

}

**SoftAssertionsAttribute.cs**
[SoftAssertionsAttribute.zip](https://github.com/nunit/nunit/files/1176666/SoftAssertionsAttribute.zip)

**Comment by CharliePoole** [comment]
This is a result of our fixing an error described in issue #652.

We treated the behavior you relied on as an error because:
1. AfterTest should run after the test
2. TearDown is traditionally considered as a part of the test.

However, in the (unimplemented) spec for #652, we indicated that we might treat attributes __located on the fixture__ differently in the future. That's actualy your specific case, so I'll leave this issue here as an enhancement request.

It's not sure that we will actually implement the enhancement, however, since we have been treating action attributes as a legacy feature. There's a chance, in other words, that a new feature will replace action attributes at some point and the enhancement you are after will be done there.

It's really a bummer that you came to rely on this and we then treated it as a bug to fix! Your best workaround is to call a method that implements the checks you need to make either at the end of each test or the start of teardown.

**Comment by andrewlaser** [comment]
@CharliePoole 
Thanks for a quick response.
I cannot run my code inside teardown because it contains Assert.Fail("xxxx");
Is there any way to run a piece of code AFTER the test code, but BEFORE TearDown?

**Comment by CharliePoole** [comment]
@andrewlaser I'm afraid not. At one point we talked about having PreTest and PostTest attributes, which would provide a common part of the test. PostTest is needed, as your use case illustrates and PreTest would be just for symmetry and to cleanly separate SetUp from non-SetUp code. The idea was never acted on. See issue #49.

**Comment by andrei-turavets-epam** [comment]
We also use custom action attribute to be run right after test and before tear down to make screenshots in case of test failures. And if this screeenshot action will be run after teardown the screenshot will not be taken in the moment of test failure. Any idea how can it be done with current situation?

**Comment by rprouse** [comment]
@andrei-ts can't you take the screenshot as the first action of the TearDown if the test fails instead of using a custom attribute? It is potentially one line in a teardown or one line as an attribute.

**Comment by andrei-ts** [comment]
@rprouse, before 3.7 it was one line in one place over our base test class which is nice reusable solution, with new update it should be one line in each test class, and as for I don't think it is right when the person who is implementing tests should take care about other things than test logic

**Comment by rprouse** [comment]
If it was one line on your base class, can't it be added to the TearDown in the base class? Just trying to provide work around as I don't think this will be fixed for 3.8.

**Comment by andrei-ts** [comment]
@rprouse, I don't think so as the parent teardown will be launched only after child teardown will be completed and I will have following chain:
TestMethod=>TestTeardown=>Parent TearDown=>CustomAction that makes screenshots
instead of
TestMethod=>CustomAction that makes screenshots in correct place =>TestTeardown=>Parent TearDown.
So if no easy workaround I will skip 3.7+ updates for our solution as of now, but it would be great if such behavior like PostTest attribute or some configurable test actions attributes will be implement in future versions

**Comment by rprouse** [comment]
You are right if you are using child TearDown and they will effect your tests. I am going to boost the priority on this a bit and see what we can come up with.

**Comment by vitali-sonchyk-epam** [comment]
Are there any updates related to the issue?
We experience the same inconvinience related to screenshot issue.

**Comment by rprouse** [comment]
@vitali-sonchyk-epam nobody has looked at this issue yet, sorry.
